File: python/2019/day_03.py
import itertools as it
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


@dataclass
class Line:
    x0: int
    y0: int
    x1: int
    y1: int
    length: int
    prev: 'Line' = None

    def __post_init__(self):
        if self.prev:
            self.length += self.prev.length


@dataclass
class Intersect:
    x: int
    y: int
    v: Line
    h: Line
    dist: int = field(init=False)
    cost: int = field(init=False)

    def __post_init__(self):
        self.dist = abs(self.x) + abs(self.y)
        self.cost = (abs(self.y - self.v.y0) + self.v.prev.length
                   + abs(self.x - self.h.x0) + self.h.prev.length)

    @staticmethod
    def try_from_lines(v: Line, h: Line) -> Optional['Intersect']:
        x0, x1 = sorted([h.x0, h.x1])
        y0, y1 = sorted([v.y0, v.y1])
        if x0 < v.x0 < x1 and y0 < h.y0 < y1:
            return Intersect(v.x0, h.y0, v, h)


@dataclass
class Wire:
    verticals: list[Line] = field(default_factory=list)
    horizontals: list[Line] = field(default_factory=list)


def main():
    text = (Path(__file__).parent / "../../input/2019/input_03.txt").read_text()
    wires = []
    # Loop over each wire path
    for path in text.splitlines():
        wires.append(wire := Wire())
        prev = None
        x1, y1 = 0, 0
        # Evaluate each move
        for move in path.strip().split(','):
            x0, y0 = x1, y1
            way, dist = move[0], int(move[1:])
            if way == 'U':
                y1 -= dist
            elif way == 'R':
                x1 += dist
            elif way == 'D':
                y1 += dist
            elif way == 'L':
                x1 -= dist
            line = Line(x0, y0, x1, y1, dist, prev)
            prev = line
            if way in {'U', 'D'}:
                wire.verticals.append(line)
            else:
                wire.horizontals.append(line)
    # Compute all intersections
    intersections = []
    for a, b in [wires, reversed(wires)]:
        for v, h in it.product(a.verticals, b.horizontals):
            if intersection := Intersect.try_from_lines(v, h):
                intersections.append(intersection)
    # Part 1
    nearest = min(intersections, key=lambda i: i.dist)
    print("Part 1:", nearest.dist)
    # Part 2
    best = min(intersections, key=lambda i: i.cost)
    print("Part 2:", best.cost)


if __name__ == "__main__":
    main()

File: rust/2019/src/bin/day_03.rs
use std::fs;

#[derive(Clone, Copy)]
struct Line {
    start: (i32, i32),
    end: (i32, i32),
    prev: i32,
}

struct Wire {
    horis: Vec<Line>,
    verts: Vec<Line>,
}

struct Intersect {
    x: i32,
    y: i32,
    v: Line,
    h: Line,
}

impl Intersect {
    fn dist(&self) -> i32 {
        self.x.abs() + self.y.abs()
    }

    fn cost(&self) -> i32 {
        (self.x - self.h.start.0).abs()
            + self.h.prev
            + (self.y - self.v.start.1).abs()
            + self.v.prev
    }

    fn find(v: &Line, h: &Line) -> Option<Intersect> {
        let (x0, x1) = if h.start.0 < h.end.0 {
            (h.start.0, h.end.0)
        } else {
            (h.end.0, h.start.0)
        };
        let (y0, y1) = if v.start.1 < v.end.1 {
            (v.start.1, v.end.1)
        } else {
            (v.end.1, v.start.1)
        };
        match (x0..x1).contains(&v.start.0) && (y0..y1).contains(&h.start.1) {
            true => Some(Intersect {
                x: v.start.0,
                y: h.start.1,
                v: *v,
                h: *h,
            }),
            false => None,
        }
    }
}

fn main() {
    let text = fs::read_to_string("../input/2019/input_03.txt").unwrap();
    // Parse wire paths from input file
    let wires: Vec<_> = text
        .lines()
        .map(|line| {
            let mut wire = Wire {
                horis: vec![],
                verts: vec![],
            };
            let (mut x1, mut y1) = (0, 0);
            let mut prev = 0;
            for step in line.split(',') {
                let (x0, y0) = (x1, y1);
                let length: i32 = step[1..].parse().unwrap();
                match &step[..1] {
                    "U" => y1 -= length,
                    "R" => x1 += length,
                    "D" => y1 += length,
                    "L" => x1 -= length,
                    s => panic!("Unknown direction: {}", s),
                }
                let line = Line {
                    start: (x0, y0),
                    end: (x1, y1),
                    prev,
                };
                match &step[..1] {
                    "U" | "D" => wire.verts.push(line),
                    "L" | "R" => wire.horis.push(line),
                    _ => unreachable!(),
                }
                prev += length;
            }
            wire
        })
        .collect();
    // Find all wire intersections
    let mut intersections = vec![];
    for (a, b) in [(&wires[0], &wires[1]), (&wires[1], &wires[0])].iter() {
        for v in &a.verts {
            for h in &b.horis {
                if let Some(intersect) = Intersect::find(v, h) {
                    intersections.push(intersect);
                }
            }
        }
    }
    println!(
        "Part 1: {}",
        intersections.iter().map(|i| i.dist()).min().unwrap()
    );
    println!(
        "Part 2: {}",
        intersections.iter().map(|i| i.cost()).min().unwrap()
    );
}
